alf/bin/train.py

```diff
@@ -67,54 +71,149 @@ def _define_flags():
+
+        if trainer_conf.ml_type == 'rl':
+            trainer = policy_trainer.RLTrainer(trainer_conf)
+        elif trainer_conf.ml_type == 'sl':
+            # NOTE: SLTrainer does not support distributed training yet
+            if world_size > 1:
+                raise RuntimeError(
+                    "Multi-GPU DDP training does not support supervised learning"
+                )
+            trainer = policy_trainer.SLTrainer(trainer_conf)
+        else:
+            raise ValueError("Unsupported ml_type: %s" % trainer_conf.ml_type)
+
+        trainer.train()
+
+        # Load the configuration file, note that this will also bring up the
+    except Exception as e:
+        # If the training worker is running as a process in multiprocessing
+        # environment, this will make sure that the exception raised in this
+        # particular process is captured and shown.
+        tb = traceback.format_exc()
+        print(e)
+        print(tb)
+    finally:
+        # Note that each training worker will have its own child processes
+        # running the environments. In the case when training worker process
+        # finishes ealier (e.g. when it raises an exception), it will hang
+        # instead of quitting unless all child processes are killed.
+        alf.close_env()
 
 
 def main(_):
-    FLAGS.alsologtostderr = True
     root_dir = common.abs_path(FLAGS.root_dir)
     os.makedirs(root_dir, exist_ok=True)
-    logging.get_absl_handler().use_absl_log_file(log_dir=root_dir)
 
     if FLAGS.store_snapshot:
-        # ../<ALF_REPO>/alf/bin/train.py
-        file_path = os.path.abspath(__file__)
-        alf_root = str(pathlib.Path(file_path).parent.parent.parent.absolute())
-        # generate a snapshot of ALF repo as ``<root_dir>/alf``
-        common.generate_alf_root_snapshot(alf_root, root_dir)
+        _snapshot_alf(root_dir)
 
     conf_file = common.get_conf_file()
-    try:
-        common.parse_conf_file(conf_file)
-        train_eval(root_dir)
-    finally:
-        alf.close_env()
+
+    # FLAGS.distributed is guaranteed to be one of the possible values.
+    if FLAGS.distributed == 'none':
+        training_worker(
+            rank=0, world_size=1, conf_file=conf_file, root_dir=root_dir)
+    elif FLAGS.distributed == 'multi-gpu':
+        world_size = torch.cuda.device_count()
+
+        if world_size == 1:
+            logging.warn(
+                'Fallback to single GPU mode as there is only one GPU')
+            training_worker(
+                rank=0, world_size=1, conf_file=conf_file, root_dir=root_dir)
+            return
+
+        # Force setting multiprocessing's start method to spawn. This is
+        # required for DDP to run correctly.
+        try:
+            mp.set_start_method('spawn', force=True)
+        except RuntimeError:
+            pass
+
+        processes = []
+        for i in range(world_size):
+            processes.append(
```

Comment:
why not use mp.spawn?

Reply:
Because I did not know about `mp.spawn` ... Thanks for mentioning that. Rewrite this with `mp.spawn` and it looks much cleaner and nicer.